samples/browser/pub_sub/index.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+const iotsdk = require("aws-iot-device-sdk-v2");
+const iot = iotsdk.iot;
+const mqtt = iotsdk.mqtt;
+const AWS = require("aws-sdk");
+const Settings = require("./Settings");
+const $ = require("jquery");
+
+function log(msg) {
+  $("#console").append(`<pre>${msg}</pre>`);
+}
+
+async function fetch_credentials() {
+  return new Promise(async (resolve, reject) => {
+    AWS.config.region = Settings.AWS_REGION;
+    if (
+      Settings.AWS_COGNITO_IDENTITY_POOL_ID !=
+      "Optional: <your identity pool id>"
+    ) {
+      const credentials = (AWS.config.credentials =
+        new AWS.CognitoIdentityCredentials({
+          IdentityPoolId: Settings.AWS_COGNITO_IDENTITY_POOL_ID,
+        }));
+      log("Fetching Cognito credentials");
+      credentials.refresh((err) => {
+        if (err) {
+          log(`Error fetching cognito credentials: ${err}`);
+          reject(`Error fetching cognito credentials: ${err}`);
+        }
+        log("Cognito credentials refreshed");
+        resolve(credentials);
+      });
+    } else {
+      const credentials = (AWS.config.credentials = new AWS.Credentials({
+        accessKeyId: Settings.AWS_STATIC_ACCESS_KEY,
+        secretAccessKey: Settings.AWS_STATIC_SECRET_ACCESS_KEY,
+        sessionToken:
+          Settings.AWS_STATIC_ACCESS_TOKEN ==
+          "Optional: <your static access token>"
+            ? undefined
+            : Settings.AWS_STATIC_ACCESS_TOKEN,
+      }));
+      resolve(credentials);
+    }
+  });
+}
+
+async function connect_websocket(credentials) {
+  return new Promise((resolve, reject) => {
+    let config =
+      iot.AwsIotMqttConnectionConfigBuilder.new_builder_for_websocket()
+        .with_clean_session(true)
+        .with_client_id(`pub_sub_sample(${new Date()})`)
+        .with_endpoint(Settings.AWS_IOT_ENDPOINT)
+        .with_credentials(
```

Comment:
Is it possible to update this to fetch the credentials as part of the signing handshake async flow?  Cognito credentials are session-based which means this sample will eventually go into a permanent failure mode on the first reconnect after session expiration.

Reply:
We're already using aws-sdk-js in this sample.

Reply:
Revisiting this, I think my comment is being misinterpreted.

As it stands, this sample (conditionally) uses Cognito but the sourced credentials are "static" and so someone taking this sample directly will not have a good starting point: as soon as the connection drops the first time after credential expiration, every reconnect will fail.  It's slightly more complicated, but if the sample instead has a handshake transform callback that sources credentials and then signs with the sourced credentials, the sample will not suffer from this defect and people using it as a starting point won't run into this fundamental issue.

Reply:
I don't think the URL transform is exposed as well. I agree the current sample for Cognito is not really helpful. What about we push the sample with static credentials only first to unblock some browser customers and expose either the credential provider override or the URL transform override from crt later?